### PR TITLE
convert S3 compute log manager to support new captured log API

### PR DESCRIPTION
**Pull request by prha**
### Summary & Motivation
This PR changes the S3 compute log manager to support the captured log manager API.  It also uses those APIs to add functionality to support streaming uploads.  

The run worker instance of the S3 compute log manager will periodically upload the in-progress compute logs to an S3 bucket using a "partial" file extension.  It does this by spawning a process that uploads the logs and sleeps, in a loop.  The compute log manager on the dagit instance now is aware of partial files in the S3 bucket, downloads them, and displays the partial data.

### How I Tested These Changes
BK, manual testing, added new test that checks for partial uploads.

### CODE REVIEW

**Comment by prha** [comment]
Current dependencies on/for this PR:
* master
  * **PR #9429** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9429" target="_blank">Graphite</a> 
    * **PR #9956** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9956" target="_blank">Graphite</a> 
      * **PR #9957** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9957" target="_blank">Graphite</a> 
        * **PR #9958** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9958" target="_blank">Graphite</a> 
          * **PR #9959** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9959" target="_blank">Graphite</a> 
            * **PR #9960** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9960" target="_blank">Graphite</a> 
              * **PR #9961** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9961" target="_blank">Graphite</a>  👈
                * **PR #9962** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9962" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/9961?utm_source=stack-comment).

File: python_modules/dagster/dagster/_grpc/server.py [hunk 1]
```diff
@@ -327,6 +334,36 @@ def _get_repo_for_origin(
             )
         return loaded_repos.definitions_by_name[external_repo_origin.repository_name]
 
+    def _compute_log_thread(self):
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_grpc/server.py:337]: definitely sketched out by this - seems like there has to be a better way

File: python_modules/libraries/dagster-aws/dagster_aws/s3/poll_upload.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        import psutil  # pylint: disable=import-error
+
+        try:
+            parent = psutil.Process(parent_pid)
+            return parent.status() != psutil.STATUS_RUNNING
+        except psutil.NoSuchProcess:
+            return True
+
+    else:
+        return os.getppid() != parent_pid
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/poll_upload.py:22]: do we not have a similar thing defined in a more core place? I know we got an external PR to fix-up similar code in the past

File: python_modules/libraries/dagster-aws/dagster_aws/s3/poll_upload.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    compute_log_manager = deserialize_json_to_dagster_namedtuple(args[1]).rehydrate()
+    log_key = json.loads(args[2])
+    interval = int(args[3])
+
+    while not compute_log_manager.is_capture_complete(log_key):
+        if pop_captured_interrupt() or (parent_pid and current_process_is_orphaned(parent_pid)):
+            return
+
+        time.sleep(interval)
+        compute_log_manager.on_progress(log_key)
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/poll_upload.py:39]: sending over the manager and rehydrating it to do this feels odd to me. This path is under the control of `S3ComputeLogManager` so going back through a generic `on_progress` interface doesn't seem necessary.

File: python_modules/dagster/dagster/_utils/__init__.py [hunk 4]
```diff
@@ -371,12 +371,14 @@ def ensure_dir(file_path):
...
         if e.errno != errno.EEXIST:
             raise
+    return file_path
 
 
 def ensure_file(path):
     ensure_dir(os.path.dirname(path))
     if not os.path.exists(path):
         touch_file(path)
+    return path
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_utils/__init__.py:381]: nit: type these fns

Comment (alangenfeld) [inline python_modules/dagster/dagster/_utils/__init__.py:381]: nit: type these

File: python_modules/libraries/dagster-aws/dagster_aws/s3/compute_log_manager.py [hunk 5]
```diff
@@ -107,109 +118,375 @@ def config_type(cls):
...
+    def __init__(self, manager):
+        self._manager = manager
+        self._subscriptions = defaultdict(list)
+        self.__polling_thread = threading.Thread(
+            target=self._poll,
+            name="s3-compute-log-streaming",
+        )
+        self.__shutdown_event = threading.Event()
+        self.__polling_thread.daemon = True
+        self.__polling_thread.start()
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/compute_log_manager.py:411]: nit: single `_` prefix

**Review by alangenfeld — APPROVED** [review]
sorry for the delay here

**Pull request merged**